pkg/transport/custom_transform_collection.go

```diff
@@ -0,0 +1,243 @@
+	// be removed from groupResourceTransformData.bindingsThatCare.
+	// No Set[GroupResource] here is empty.
+	bindingNameToGroupResources map[string]sets.Set[metav1.GroupResource]
+}
+
+// groupResourceTransformData is the ingested custom transforms for a given GroupResource
+type groupResourceTransformData struct {
+	bindingsThatCare sets.Set[string /*Binding name*/] // not empty
+	removes          []jsonpath.Query
+}
+
+func newCustomTransformCollection(client controlclient.CustomTransformInterface, getTransformObjects func(indexName, indexedValue string) ([]any, error), enqueue func(any)) *customTransformCollection {
+	return &customTransformCollection{
+		client:                      client,
+		getTransformObjects:         getTransformObjects,
+		enqueue:                     enqueue,
+		grToTransformData:           make(map[metav1.GroupResource]*groupResourceTransformData),
+		ctNameToSpec:                make(map[string]v1alpha1.CustomTransformSpec),
+		bindingNameToGroupResources: make(map[string]sets.Set[metav1.GroupResource]),
+	}
+}
+
+// GetCustomTransformData returns the groupResourceTransformData to use
+// for the given GroupResource on behalf of the named Binding.
+// This method returns a cached answer if one is available, otherwise
+// digests the relevant CustomTransform object(s) and caches the result.
+// Always records the fact that the given binding depends on the answer.
+func (ctc *customTransformCollection) GetCustomTransformData(ctx context.Context, groupResource metav1.GroupResource, bindingName string) *groupResourceTransformData {
+	logger := klog.FromContext(ctx)
+	ctc.mutex.Lock()
+	defer ctc.mutex.Unlock()
+	grTransformData, ok := ctc.grToTransformData[groupResource]
+	if !ok {
+		grTransformData = &groupResourceTransformData{
+			bindingsThatCare: sets.New(bindingName),
+		}
+		ctKey := customTransformDomainKey(groupResource.Group, groupResource.Resource)
+		ctAnys, err := ctc.getTransformObjects(customTransformDomainIndexName, ctKey)
+		if err != nil {
+			// This only happens if the index is not defined;
+			// that is, it never happens.
+			// If it does, retry will not help.
+			logger.Error(err, "Failed to get objects from CustomTransform domain index", "key", ctKey)
+		}
+		ctNames := []string{}
+
+		// Digest each relevant CustomTransform, accumulating remove instructions in groupResourceTransforData.removes.
+		// Invalidate cache entry for each CustomTransform that changed its Spec's .Group or .Resource.
+		for _, ctAny := range ctAnys {
+			ct := ctAny.(*v1alpha1.CustomTransform)
+			removes := ctc.digestCustomTransformLocked(ctx, groupResource, bindingName, ct)
+			grTransformData.removes = append(grTransformData.removes, removes...)
+			ctNames = append(ctNames, ct.Name)
+		}
+		if len(ctAnys) > 1 { // This is not recommended
+			logger.Error(nil, "Multiple CustomTransform objects apply to one GroupResource", "groupResource", groupResource, "names", ctNames)
```

Comment:
This does not follow the use of Error across the rest of the codebase. 

Additionally, perhaps this is something we should think of. E.g., for the following: https://github.com/kubestellar/kubestellar/issues/2064#issuecomment-2067992923

Karmada has a similar feature as well, but they implement it through 2 different resources.

Reply:
@vMaroon:

I am not sure I get your meaning. If your point about this Error log statement is that it is unusual to use a log statement to tell the user that they are doing something not recommended, then I agree. This message should appear in an API object. Since it is not a fatal error, I did not make the controller put it in the `status.errors` of the Binding or CustomTransform. I think that what we should do is introduce `Warnings []` in CustomTransformStatus and report it there. I was planning to get to that after this PR merges. I also wanted to have some discussion about whether we are really agreed that this is not recommended (and would say so in the to-be-written doc if agreed).

Reply:
The point was this use is unusual, yes. I guess it can be left until the API is more mature to contain errors. Do we really want warnings in the API? What qualifies a warning? A state that is fine but can go wrong? Or what?

The 2nd part of my comment refers to `if len(ctAnys) > 1 { // This is not recommended` - the whole matter. Perhaps now is not the time for it, but I think it's not insensible for use-cases of chaining transformations (in order) to appear, such as what's mentioned in the comment. Therefore if possible, the solution should not be fragile to the idea (should not support it, but should not be fragile - which it's not).

Reply:
So the first question is whether we want to recommend that restriction.  I have opened Issue #2067 about that.